src/hotspot/share/prims/jvmtiThreadState.cpp

```diff
@@ -258,11 +281,45 @@ JvmtiVTMSTransitionDisabler::~JvmtiVTMSTransitionDisabler() {
   if (Thread::current_or_null() == NULL) {
     return;  // Detached thread, can be a call from Agent_OnLoad.
   }
-  enable_VTMS_transitions();
+  if (_vthread != NULL) {
+    VTMS_transition_enable_for_one(); // enable VTMS transitions for one virtual thread
+  } else {
+    VTMS_transition_enable_for_all(); // enable VTMS transitions for all virtual threads
+  }
+}
+
+// disable VTMS transitions for one virtual thread
+void
+JvmtiVTMSTransitionDisabler::VTMS_transition_disable_for_one() {
+  assert(_vthread != NULL, "sanity check");
+  JavaThread* thread = JavaThread::current();
+  HandleMark hm(thread);
+  Handle vth = Handle(thread, JNIHandles::resolve_external_guard(_vthread));
+  if (!java_lang_VirtualThread::is_instance(vth())) {
+    return;
+  }
+  JvmtiThreadState* vstate = java_lang_Thread::jvmti_thread_state(vth());
+
+  ThreadBlockInVM tbivm(thread);
+  MonitorLocker ml(JvmtiVTMSTransition_lock, Mutex::_no_safepoint_check_flag);
```

Comment:
Aside: this pattern looks very odd. Why not just lock with the safepoint check?

Reply:
@dholmes-ora
Thank you for looking at this fix and for the comments.
I will on replying and resolving them.

Reply:
This update just follows the common pattern which was introduced about two years ago. At the moment I do not remember the exact reason. This code needs to wait for the counter to become zero and to use `ThreadBlockInVM` which can reach a safepoint in the destructor. Can we lock with a safepoint check in this case? Or we should get rid of the `ThreadBlockInVM` then?
I'll check if this can be fixed.

Reply:
This can be a separate RFE if the reason for it being a nosafepoint lock no longer holds. It would be cleaner to use a normal safepoint checking lock than to artificially make the thread safepoint-safe before using it.

Reply:
Agreed. I was thinking about a separate RFE. I started seeing some failures in the S/R related tests with the removed `ThreadBlockInVM` and `_no_safepoint_check_flag`. Need to investigated why. Otherwise, it seems to be working well.

Reply:
As I see there can be some problems with the JVMTI Suspend/Resume but it is worth to investigate anyway.
So, I've filed an RFE:
[JDK-8299240](https://bugs.openjdk.org/browse/JDK-8299240): rank of JvmtiVTMSTransition_lock can be safepoint